src/hdmf/data_utils.py

```diff
@@ -922,7 +922,7 @@ class DataIO:
     def __init__(self, **kwargs):
         data, dtype, shape = popargs('data', 'dtype', 'shape', kwargs)
         if data is None:
-            if dtype is None or shape is None:
+            if (dtype is None) ^ (shape is None):
```

Comment:
Why use XOR instead of OR here? Should this error not also be raised if both dtype and shape are None?

Reply:
See the commit message. It's done for backward compatibility i.e. set_dataio